### PR TITLE
Refactor transaction subsystem to support failover for transactions

### PR DESCRIPTION
**Pull request by kuujo**
* Replicate participant info in TransactionService
* When a member leaves the cluster, use atomic operations to assume control over its transactions
* Commit or roll back transactions based on transaction state

This needs a lot more testing, but I'm throwing it up here now for review.

### CODE REVIEW

**Comment by kuujo** [comment]
#360

File: core/src/main/java/io/atomix/core/impl/CoreTransactionService.java [hunk 1]
```diff
@@ -99,13 +175,190 @@ public TransactionState getTransactionState(TransactionId transactionId) {
...
+  /**
+   * Completes a transaction in the {@link TransactionState#PREPARING} state.
+   *
+   * @param transactionId the transaction identifier for the transaction to complete
+   */
+  private void completePreparingTransaction(TransactionId transactionId) {
+    completeTransaction(
+        transactionId,
+        TransactionState.PREPARING,
+        info -> new TransactionInfo(localMemberId, TransactionState.ROLLING_BACK, info.participants),
```
Comment (johnou) [inline core/src/main/java/io/atomix/core/impl/CoreTransactionService.java]: PREPARING?

File: core/src/main/java/io/atomix/core/impl/CoreTransactionService.java [hunk 2]
```diff
@@ -99,13 +175,190 @@ public TransactionState getTransactionState(TransactionId transactionId) {
...
+        info -> info.state == TransactionState.ROLLING_BACK,
```
Comment (johnou) [inline core/src/main/java/io/atomix/core/impl/CoreTransactionService.java]: PREPARING?

File: core/src/main/java/io/atomix/core/impl/CoreTransactionService.java (hunk 1 above)
Comment (johnou) [reply core/src/main/java/io/atomix/core/impl/CoreTransactionService.java]: ah no, missed transactional.rollback below

File: core/src/main/java/io/atomix/core/impl/CoreTransactionService.java [hunk 3]
```diff
@@ -99,13 +175,190 @@ public TransactionState getTransactionState(TransactionId transactionId) {
...
+
+  /**
+   * Completes an individual participant in a transaction by loading the primitive by type/protocol/partition group
+   * and applying the given completion function to it.
+   */
+  @SuppressWarnings("unchecked")
+  private CompletableFuture<Void> completeParticipant(
+      ParticipantInfo participantInfo,
+      Function<Transactional<?>, CompletableFuture<Void>> completionFunction) {
+    // Loop up the primitive type for the participant. If the primitive type is not found, return an exception.
```
Comment (johnou) [inline core/src/main/java/io/atomix/core/impl/CoreTransactionService.java]: look up?

**Pull request merged**